src/btt.c

```diff
@@ -801,6 +833,58 @@ read_arenas(struct btt *bttp, int lane, int narena)
 	return -1;
 }
 
+static int
+write_map(struct btt *bttp, int lane, off_t map_entry_off,
+		uint64_t external_nlba)
+{
+	/* write out the initial map, identity style */
+	uint32_t *mapp = NULL;
+	ssize_t mlen = 0;
+	int next_index = 0;
+	ssize_t remaining = 0;
+	for (int i = 0; i < external_nlba; i++) {
+		if (remaining == 0) {
+			/* flush previous mapped area */
+			if (mapp != NULL) {
+			/*
+			 * Protect the memory again
+			 * (debug version only).
+			 * If (mapp != NULL) it had to be
+			 * unprotected earlier.
+			 */
+				RANGE_RO(mapp, mlen);
+
+				(*bttp->ns_cbp->nssync)(bttp->ns,
+						lane, mapp, mlen);
+			}
+			/* request a mapping of remaining map area */
+			mlen = (*bttp->ns_cbp->nsmap)(bttp->ns,
+					lane, (void **) &mapp,
+					(external_nlba - i) * sizeof (uint32_t),
+					map_entry_off);
+
+			if (mlen < 0)
+				return -1;
+
+			/* unprotect the memory (debug version only) */
+			RANGE_RW(mapp, mlen);
+
+			remaining = mlen;
+			next_index = 0;
+		}
+		mapp[next_index++] = htole32(i | BTT_MAP_ENTRY_ZERO);
```

Comment:
I think that it’s a good idea. I’ll use the pmem_memset for the nszero callback in pmemblk. The new sse versions written by Sarah should be pretty good for this.

From: Andy Rudoff [mailto:notifications@github.com]
Sent: Saturday, January 17, 2015 7:15 PM
To: pmem/nvml
Cc: Kapela, Tomasz
Subject: Re: [nvml] blk: change how the EZ flags are handled in the blk map (#109)


In src/btt.c<https://github.com/pmem/nvml/pull/109#discussion-diff-23127221>:

> +                    /* request a mapping of remaining map area */

> +                    mlen = (*bttp->ns_cbp->nsmap)(bttp->ns,

> +                                   lane, (void **) &mapp,

> +                                   (external_nlba - i) * sizeof (uint32_t),

> +                                   map_entry_off);

> +

> +                    if (mlen < 0)

> +                            return -1;

> +

> +                    /* unprotect the memory (debug version only) */

> +                    RANGE_RW(mapp, mlen);

> +

> +                    remaining = mlen;

> +                    next_index = 0;

> +            }

> +            mapp[next_index++] = htole32(i | BTT_MAP_ENTRY_ZERO);

My intention (which I realize I didn't state very clearly) was that the map either known to be initialized to zero, or you initialize it to zero. I think that write_map() no longer needs to store the identity values because you assume the value is the identity when btt_is_initial() returns true.

So I like the logic you have, I just think write_map() should simply zero the map.

Given this, I wonder if we should introduce an nszero callback so the caller can use whatever optimized methods are available for zeroing pmem. What do you think?

—
Reply to this email directly or view it on GitHub<https://github.com/pmem/nvml/pull/109/files#r23127221>.
--------------------------------------------------------------------

Intel Technology Poland sp. z o.o.
ul. Slowackiego 173 | 80-298 Gdansk | Sad Rejonowy Gdansk Polnoc | VII Wydzial Gospodarczy Krajowego Rejestru Sadowego - KRS 101882 | NIP 957-07-52-316 | Kapital zakladowy 200.000 PLN.

Ta wiadomosc wraz z zalacznikami jest przeznaczona dla okreslonego adresata i moze zawierac informacje poufne. W razie przypadkowego otrzymania tej wiadomosci, prosimy o powiadomienie nadawcy oraz trwale jej usuniecie; jakiekolwiek
przegladanie lub rozpowszechnianie jest zabronione.
This e-mail and any attachments may contain confidential material for the sole use of the intended recipient(s). If you are not the intended recipient, please contact the sender and delete all copies; any review or distribution by
others is strictly prohibited.